vortex-scan/src/multi_scan.rs

```diff
@@ -46,44 +56,65 @@ impl<T> MultiScan<T> {
     ///
     /// The scan progresses when calling `next` on the iterator.
     pub fn new_scan_iterator(&self) -> MultiScanIterator<T> {
+        let worker = Worker::new_fifo();
+        self.stealers.insert(
+            self.next_construction_id.fetch_add(1, SeqCst),
+            worker.stealer(),
+        );
+
         MultiScanIterator {
             scan_builder_factory: self.scan_builder_factory.clone(),
             local_pool: LocalPool::new(),
-            task_queue: SegQueue::new(),
+            stealers: self.stealers.clone(),
+            next_stealer_id: self.next_stealer_id.clone(),
+            worker,
         }
     }
 }
 
 /// Scan iterator to participate in a `MultiScan`.
 pub struct MultiScanIterator<T> {
     local_pool: LocalPool,
+    worker: Worker<ArrayFuture<T>>,
+    stealers: Arc<DashMap<usize, Stealer<ArrayFuture<T>>>>,
+    next_stealer_id: Arc<AtomicUsize>,
 
     /// Thread-safe queue of closures that lazily produce [`ScanBuilder`] instances.
     /// This queue is shared across all iterators being created with `new_scan_iterator`.
     scan_builder_factory: ScanBuilderFactory<T>,
-    task_queue: SegQueue<ArrayFuture<T>>,
 }
 
 impl<T: Send + Sync + 'static> Iterator for MultiScanIterator<T> {
     type Item = VortexResult<T>;
 
     fn next(&mut self) -> Option<VortexResult<T>> {
         // Queue up tasks if the thread local queue is empty.
-        if self.task_queue.is_empty() {
+        if self.worker.is_empty() {
             if let Some(scan_builder_fn) = self.scan_builder_factory.pop() {
                 match scan_builder_fn().build() {
                     Ok(tasks) => {
                         for task in tasks {
-                            self.task_queue.push(Box::pin(task));
+                            self.worker.push(Box::pin(task));
                         }
                     }
                     Err(err) => return Some(Err(err)),
                 }
+            } else {
+                for _ in 0..self.stealers.len() {
+                    // Round robin to ensure work is not always stolen from the same worker.
+                    let steal_id = self.next_stealer_id.fetch_add(1, SeqCst) % self.stealers.len();
+                    if let Some(stealer) = self.stealers.get(&steal_id) {
+                        if !stealer.is_empty() {
+                            // Steal ~half of the work and push it into `worker`.
+                            _ = stealer.steal_batch(&self.worker);
+                            break;
+                        }
+                    }
+                }
             }
-            // TODO(Alex): worksteal tasks from other threads
         }
 
-        let task = self.task_queue.pop()?;
+        let task = self.worker.pop()?;
 
         self.local_pool.run_until(task).transpose()
```

Comment:
We should figure out next how to run ~2 tasks on the pool, rather than just one. That way we can do CPU while waiting for I/O

Reply:
We can just push n futures into `FuturesUnordered` and poll on that.